FileStorageLevels/ReadLeveledStorage.cpp

```diff
@@ -0,0 +1,309 @@
+    }
+    if(position == 0 && extents.begin()->offset == 0 && extents.begin()->length == 0)
+    {
+        // delete convenience placeholder, since we have real data to populate it with.
+        extents.erase(extents.begin());
+    }
+
+    auto emplaced = extents.emplace(Extent{.offset = position, .length = static_cast<size_t>(write_res)});
+    if(!emplaced.second)
+    {
+        throw std::logic_error("Failed to emplace new extent - offset: " + std::to_string(emplaced.first->offset) +
+                               ", length: " + std::to_string(emplaced.first->length));
+    }
+    mergeAdjacent(emplaced.first);
+    position += write_res;
+    primed_file_size += write_res;
+    is_fully_primed = primed_file_size == file_size;
+    if(primed_file_size > file_size)
+    {
+        throw std::logic_error("Current primary file size " + std::to_string(primed_file_size) + " is greater than secondary file size " +
+                               std::to_string(file_size));
+    }
+    if(is_fully_primed)
+    {
+        extents.clear();
+        extents.emplace(Extent{.offset = 0, .length = file_size});
+    }
+    return write_res;
+}
+
+void ReadLeveledStorage::mergeAdjacent(Extents::const_iterator it)
+{
+    validateAdjacent(it);
+    {
+        auto prev = std::prev(it);
+        if(prev != it && prev->offset + prev->length == it->offset)
+        {
+            Extent new_extent;
+            new_extent.offset = prev->offset;
+            new_extent.length = prev->length + it->length;
+            new_extent.merge_level = prev->merge_level + it->merge_level + 1;
+            extents.erase(prev);
+            extents.erase(it);
+            auto emplaced = extents.emplace(new_extent);
+            if(!emplaced.second)
+            {
+                throw std::logic_error("Failed to emplace new extent - offset: " + std::to_string(emplaced.first->offset) +
+                                       ", length: " + std::to_string(emplaced.first->length));
+            }
+            it = emplaced.first;
+        }
+    }
+
+    {
+        auto next = std::next(it);
+        if(next != extents.end() && it->offset + it->length == next->offset)
+        {
+            Extent new_extent;
+            new_extent.offset = it->offset;
+            new_extent.length = it->length + next->length;
+            new_extent.merge_level = next->merge_level + it->merge_level + 1;
+            extents.erase(next);
+            extents.erase(it);
+            auto emplaced = extents.emplace(new_extent);
+            if(!emplaced.second)
+            {
+                throw std::logic_error("Failed to emplace new extent - offset: " + std::to_string(emplaced.first->offset) +
+                                       ", length: " + std::to_string(emplaced.first->length));
+            }
+        }
+    }
+}
+
+void ReadLeveledStorage::validateAdjacent(Extents::const_iterator it) const
+{
+    auto prev = std::prev(it);
```

Comment:
Could 'it' refer to extents.begin() here?

Reply:
already pushed fix